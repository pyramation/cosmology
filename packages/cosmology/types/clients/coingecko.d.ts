import { PriceHash } from '../..'
/**
 * @typedef {('cosmos'|
 * 'osmosis'|
 * 'ion'|
 * 'akash-network'|
 * 'sentinel'|
 * 'iris-network'|
 * 'crypto-com-chain'|
 * 'persistence'|
 * 'regen'|
 * 'starname'|
 * 'e-money'|
 * 'e-money-eur'|
 * 'juno-network'|
 * 'likecoin'|
 * 'terrausd'|
 * 'terra-luna'|
 * 'bitcanna'|
 * 'terra-krw'|
 * 'secret'|
 * 'medibloc'|
 * 'comdex'|
 * 'stargaze'|
 * 'chihuahua-token'|
 * 'cheqd-network'|
 * 'vidulum')} CoinGeckoToken
 *
 * @typedef {Object.<CoinGeckoToken, {usd: number}>} TokenPricesUSDResponse
 *
 */
export declare const CoinGeckoToken: {
  cosmos: string
  osmosis: string
  ion: string
  'akash-network': string
  sentinel: string
  'iris-network': string
  'crypto-com-chain': string
  persistence: string
  regen: string
  starname: string
  'e-money': string
  'e-money-eur': string
  'juno-network': string
  likecoin: string
  terrausd: string
  'terra-luna': string
  bitcanna: string
  'terra-krw': string
  secret: string
  medibloc: string
  comdex: string
  'cheqd-network': string
  vidulum: string
  stargaze: string
  'chihuahua-token': string
}
/**
 * @param {*} coins is a list of coins to check
 */
export declare const getPrices: (coins?: string[]) => Promise<any>
export declare const allGeckoAssets: () => any
export declare const _getPricesFromCoinGecko: () => Promise<any>
export declare const getPricesFromCoinGecko: () => Promise<PriceHash>
