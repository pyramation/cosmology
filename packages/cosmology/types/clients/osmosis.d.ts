import { CosmosApiClient } from './cosmos'
import { LcdPool, Pool, PoolDisplay, PoolPretty } from '../..'
export interface PoolsResponse {
  pools: LcdPool[]
  pagination: object
}
export declare class OsmosisApiClient extends CosmosApiClient {
  constructor ({ url }?: { url?: string })
  getPools (): Promise<PoolsResponse>
  getPool (poolId: any): Promise<any>
  getAccountLockedLongerDuration (address: any): Promise<any>
  getAccountLockedCoins (address: any): Promise<any>
  getEpochProvision (): Promise<any>
  getEpochs (): Promise<any>
  getDistrInfo (): Promise<any>
  getParams (): Promise<any>
  getLockableDurations (): Promise<any>
  getGauges (): Promise<any>
  getActiveGauges (): Promise<any>
  getGauge (gaugeId: any): Promise<any>
  getIncentivizedPools (): Promise<any>
  getPoolsPretty ({
    includeDetails
  }?: {
    includeDetails?: boolean
  }): Promise<PoolPretty[]>
  getPoolPretty (
    poodId: any,
    {
      includeDetails
    }?: {
      includeDetails?: boolean
    }
  ): Promise<PoolPretty>
}
export declare const prettyPool: (
  pool: LcdPool | Pool | PoolDisplay,
  {
    includeDetails
  }?: {
    includeDetails?: boolean
  }
) => PoolPretty
