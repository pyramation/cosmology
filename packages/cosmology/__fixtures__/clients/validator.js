import getPoolsFixture from '../validator/search/v1/pools/data.json';
import getTokensFixture from '../validator/tokens/v2/all/data.json';
import getPoolAprsFixture from '../validator/apr/v1/all/data.json';
import getPoolAprFixture from '../validator/apr/v1/606/data.json';
import getPairsSummaryFixture from '../validator/pairs/v1/summary/data.json';

import {
  ValidatorToken,
  CoinSymbol,
  ValidatorPoolApr,
  ValidatorPair,
  ValidatorTokenPrice
} from '../../src/types';

//  https://api-osmosis.imperator.co/swagger/#/
export class OsmosisValidatorClient {
  async getPools() {
    return getPoolsFixture;
  }

  /**
   * @returns {Promise<ValidatorToken[]>}
   */
  async getTokens() {
    return getTokensFixture;
  }

  /**
   * @param {CoinSymbol} symbol
   * @returns {Promise<ValidatorToken[]>}
   */
  async getToken(symbol) {
    console.warn('NOT IMPLEMENTED IN TESTING');
  }

  /**
   * @param {CoinSymbol} symbol
   * @returns {Promise<ValidatorTokenPrice[]>}
   */
  async getTokenPrice(symbol) {
    console.warn('NOT IMPLEMENTED IN TESTING');
  }

  /**
   * @returns {Promise<ValidatorPoolApr[]>}
   */
  async getPoolAprs() {
    return getPoolAprsFixture;
  }

  /**
   * @param {CoinSymbol} symbol
   * @returns {Promise<ValidatorPoolApr[]>}
   */
  async getPoolApr(symbol) {
    return getPoolAprFixture;
  }

  /**
   * @returns {Promise<ValidatorPair[]>}
   */
  async getPairsSummary() {
    return getPairsSummaryFixture;
  }
}
