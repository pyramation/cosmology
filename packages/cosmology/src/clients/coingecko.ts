import axios from 'axios';
import { assets } from '../assets';
import { PriceHash } from '../types';
import { convertGeckoPricesToDenomPriceHash } from '../utils/osmo/utils';

/**
 * @typedef {('cosmos'|
 * 'osmosis'|
 * 'ion'|
 * 'akash-network'|
 * 'sentinel'|
 * 'iris-network'|
 * 'crypto-com-chain'|
 * 'persistence'|
 * 'regen'|
 * 'starname'|
 * 'e-money'|
 * 'e-money-eur'|
 * 'juno-network'|
 * 'likecoin'|
 * 'terrausd'|
 * 'terra-luna'|
 * 'bitcanna'|
 * 'terra-krw'|
 * 'secret'|
 * 'medibloc'|
 * 'comdex'|
 * 'stargaze'|
 * 'chihuahua-token'|
 * 'cheqd-network'|
 * 'vidulum')} CoinGeckoToken
 *
 * @typedef {Object.<CoinGeckoToken, {usd: number}>} TokenPricesUSDResponse
 *
 */

//  https://api.coingecko.com/api/v3/simple/price?ids=osmosis,ion,cosmos,terra-luna,crypto-com-chain,terrausd,secret,juno-network,persistence,terra-krw,akash-network,regen,sentinel,iris-network,starname,e-money,e-money-eur,likecoin,bitcanna,medibloc,comdex,cheqd-network,vidulum,stargaze,chihuahua-token&vs_currencies=usd

export const CoinGeckoToken = {
  // Enum
  cosmos: 'cosmos',
  osmosis: 'osmosis',
  ion: 'ion',
  'akash-network': 'akash-network',
  sentinel: 'sentinel',
  'iris-network': 'iris-network',
  'crypto-com-chain': 'crypto-com-chain',
  persistence: 'persistence',
  regen: 'regen',
  starname: 'starname',
  'e-money': 'e-money',
  'e-money-eur': 'e-money-eur',
  'juno-network': 'juno-network',
  likecoin: 'likecoin',
  terrausd: 'terrausd',
  'terra-luna': 'terra-luna',
  bitcanna: 'bitcanna',
  'terra-krw': 'terra-krw',
  secret: 'secret',
  medibloc: 'medibloc',
  comdex: 'comdex',
  'cheqd-network': 'cheqd-network',
  vidulum: 'vidulum',
  stargaze: 'stargaze',
  'chihuahua-token': 'chihuahua-token'
};

/**
 * @param {*} coins is a list of coins to check
 */

export const getPrices = async (coins = ['osmosis']) => {
  //'https://api.coingecko.com/api/v3/simple/price?ids=cosmos,osmosis,ion,akash-network,sentinel,iris-network,crypto-com-chain,persistence,regen,starname,e-money,e-money-eur,juno-network,likecoin,terrausd,terra-luna,bitcanna,terra-krw,secret,medibloc,comdex,cheqd-network,vidulum&vs_currencies=usd'
  const fetchUrl = `https://api.coingecko.com/api/v3/simple/price?ids=${coins.join()}&vs_currencies=usd`;
  try {
    const response = await axios.get(fetchUrl);

    return response.data;
  } catch (e) {
    console.error(
      'CoinGecko API response error:',
      e.response ? e.response.data : e
    );
    return null;
  }
};

export const allGeckoAssets = () => {
  return assets
    .filter((asset) => !!asset.coingecko_id);
};

export const _getPricesFromCoinGecko = async () => {
  const geckoAssets = allGeckoAssets();
  const geckoIds = geckoAssets
    .map(asset => asset.coingecko_id);
  const prices = await getPrices(geckoIds);
  return prices;
};

export const getPricesFromCoinGecko = async (): Promise<PriceHash> => {
  const prices = await _getPricesFromCoinGecko();
  return convertGeckoPricesToDenomPriceHash(prices);
};
