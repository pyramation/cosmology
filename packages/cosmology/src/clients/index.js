export * from './osmosis';
export * from './cosmos';
export * from './validator';