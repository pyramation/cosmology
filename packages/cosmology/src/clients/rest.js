import axios from 'axios';
import retry from 'retry';

export class RestClient {
  constructor({ url }) {
    this.url = url.endsWith('/') ? url : `${url}/`;
    this._clientType = 'API';
    this.instance = axios.create({
      baseURL: this.url,
      timeout: 10000,
      headers: {}
    });
  }

  request(endpoint, opts = {}) {
    const operation = retry.operation({
      retries: 5,
      factor: 2,
      minTimeout: 1 * 1000,
      maxTimeout: 60 * 1000
    });

    return new Promise((resolve, reject) => {
      operation.attempt(async () => {
        let response;
        let err;
        try {
          response = await this.instance.get(endpoint, {
            timeout: 30000,
            ...opts
          });
        } catch (e) {
          console.log(`STATUS CODE: ${e?.response?.status}`);
          console.log(`STATUS TEXT: ${e?.response?.statusText}`);
          console.error(
            `${this._clientType} response error: ${
              e.response ? e.response.data : e
            }`
          );
          const statusNum = Number(e?.response?.status);
          if (statusNum >= 500) err = true;
          else return reject(e);
        }

        if (operation.retry(err)) {
          console.log('retrying...');
          return;
        }

        if (response && response.data) {
          resolve(response.data);
        } else {
          reject(operation.mainError());
        }
      });
    });
  }
}
