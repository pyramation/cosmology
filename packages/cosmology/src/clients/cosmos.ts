import { Coin } from '@cosmjs/amino';
import { RestClient } from './rest';

export interface Proposal {
  content: {
    title: string;
    description: string;
  }
  proposal_id: string;
  status: string;
  final_tally_result: { yes: string; abstain: string; no: string; no_with_veto: string; };
  submit_time: string;
  deposit_end_time: string;
  total_deposit: [Coin],
  voting_start_time: string;
  voting_end_time: string;
}

export interface ProposalsResponse {
  pagination?: object;
  proposals: Proposal[];
}

export class CosmosApiClient extends RestClient {
  constructor({ url }) {
    super({ url });
    this._clientType = 'Cosmos API';
  }

  async getTransaction(txHash) {
    const endpoint = `txs/${txHash}`;
    return await this.request(endpoint);
  }

  /**
   * @returns {Promise<{
   *   tx: {
   *      body: object;
   *      authInfo: object;
   *      signatures: object[];
   *   };
   *   tx_response: {
   *      height: string;
   *      txhash: string;
   *   }
   * }>}
   */
  async getCosmosTransaction(txHash) {
    const endpoint = `cosmos/tx/v1beta1/txs/${txHash}`;
    return await this.request(endpoint);
  }

  async getProposals(proposalStatus: number = 2): Promise<ProposalsResponse> {
    const endpoint = `cosmos/gov/v1beta1/proposals`;
    if (proposalStatus) {
      return await this.request(endpoint, { params: { proposal_status: proposalStatus } });
    }
    return await this.request(endpoint);
  }

  /**
   * @returns {Promise<{
   *   block_id: {
   *      hash: string;
   *      part_set_header: object;
   *   };
   *   block: {
   *      header: {
   *        version: object;
   *        chain_id: string;
   *        height: string;
   *        time: string;
   *      }
   *   }
   * }>}
   */

  async getLatestBlock() {
    const endpoint = `/cosmos/base/tendermint/v1beta1/blocks/latest`;
    return await this.request(endpoint);
  }

  // below is the subset only supported by keplr.app endpoint...

  async getBalances(address) {
    const endpoint = `bank/balances/${address}`;
    return await this.request(endpoint);
  }

  async authInfo(address) {
    const endpoint = `auth/accounts/${address}`;
    return await this.request(endpoint);
  }

  async getUnbondingDelegations(address) {
    const endpoint = `staking/delegators/${address}/unbonding_delegations`;
    return await this.request(endpoint);
  }

  async getDelegations(address) {
    const endpoint = `staking/delegators/${address}/delegations`;
    return await this.request(endpoint);
  }

  async getDelegatorValidatorInfo(delegatorAddr, validatorAddr) {
    const endpoint = `cosmos/staking/v1beta1/delegators/${delegatorAddr}/validators/${validatorAddr}`;
    return await this.request(endpoint);
  }

  async getValidatorInfo(validatorAddr) {
    const endpoint = `cosmos/staking/v1beta1/validators/${validatorAddr}`;
    return await this.request(endpoint);
  }

  async getValidators(status) {
    const endpoint = `cosmos/staking/v1beta1/validators`;
    if (status) {
      return await this.request(endpoint, { params: { status } });
    }
    return await this.request(endpoint);
  }

  async getRewards(address) {
    const endpoint = `cosmos/distribution/v1beta1/delegators/${address}/rewards`;
    return await this.request(endpoint);
  }
}
