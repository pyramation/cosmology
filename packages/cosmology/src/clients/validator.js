import { RestClient } from './rest';
import autobind from 'class-autobind';

import {
  ValidatorToken,
  CoinSymbol,
  ValidatorPoolApr,
  ValidatorPair,
  ValidatorTokenPrice
} from '../..;

//  https://api-osmosis.imperator.co/swagger/#/
export class OsmosisValidatorClient extends RestClient {
  constructor({ url = 'https://api-osmosis.imperator.co/' } = {}) {
    super({ url });
    this._clientType = 'Osmosis Validator';
    autobind(this); // React ES6 doesn't bind this -> meaning we get 'unable to read property 'request' of undefined
  }

  async getPools() {
    const endpoint = `search/v1/pools`;
    return await this.request(endpoint);
  }

  /**
   * @returns {Promise<ValidatorToken[]>}
   */
  async getTokens() {
    const endpoint = `tokens/v2/all`;
    return await this.request(endpoint);
  }

  /**
   * @param {CoinSymbol} symbol
   * @returns {Promise<ValidatorToken[]>}
   */
  async getToken(symbol) {
    const endpoint = `tokens/v2/${symbol}`;
    return await this.request(endpoint);
  }

  /**
   * @param {CoinSymbol} symbol
   * @returns {Promise<ValidatorTokenPrice[]>}
   */
  async getTokenPrice(symbol) {
    const endpoint = `tokens/v2/price/${symbol}`;
    return await this.request(endpoint);
  }

  /**
   * @returns {Promise<ValidatorPoolApr[]>}
   */
  async getPoolAprs() {
    const endpoint = `apr/v1/all`;
    return await this.request(endpoint);
  }

  /**
   * @param {CoinSymbol} symbol
   * @returns {Promise<ValidatorPoolApr[]>}
   */
  async getPoolApr(symbol) {
    const endpoint = `apr/v1/${symbol}`;
    return await this.request(endpoint);
  }

  /**
   * @returns {Promise<ValidatorPair[]>}
   */
  async getPairsSummary() {
    const endpoint = `pairs/v1/summary`;
    return await this.request(endpoint);
  }
}
