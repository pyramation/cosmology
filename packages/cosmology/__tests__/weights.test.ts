// @ts-nocheck
import poolsFixture from '../__fixtures__/lcd/osmosis/gamm/v1beta1/pools/data.json'
import geckoPricesFixtures from '../__fixtures__/coingecko/api/v3/simple/price/data.json'
import cases from 'jest-in-case'
import {
  symbolsAndDisplayValuesToCoinsArray,
  getTradesRequiredToGetBalances,
  getSwaps,
  convertGeckoPricesToDenomPriceHash,
  calculateCoinsTotalBalance,
  convertWeightsIntoCoins,
  makePoolPairs,
  makePoolsPretty
} from '../src/utils/osmo'
import { prettyPool } from '../src/clients/osmosis'
import { LcdPool } from '..'
import { Dec } from '@keplr-wallet/unit'

const prices = convertGeckoPricesToDenomPriceHash(geckoPricesFixtures)
const lcdPools: LcdPool[] = poolsFixture.pools
const prettyPools = makePoolsPretty(prices, lcdPools)
const pairs = makePoolPairs(prettyPools)
const pools = lcdPools.map(pool => prettyPool(pool))

const bank = {
  total: '9647.3',
  balances: [
    {
      // 36.63
      symbol: 'ATOM',
      amount: 10
    },
    {
      // 1.81
      symbol: 'AKT',
      amount: 100
    },
    {
      // 9.1
      symbol: 'OSMO',
      amount: 1000
    }
  ]
}

cases(
  'weights',
  async opts => {
    const balances = symbolsAndDisplayValuesToCoinsArray(opts.balances)

    const balanceValue = calculateCoinsTotalBalance({
      prices,
      coins: balances
    })
    expect(new Dec(balanceValue).sub(new Dec(opts.total)).isZero())

    const weights = opts.weights
    const weightCoins = convertWeightsIntoCoins({
      weights,
      pools,
      prices,
      balances
    })

    const poolsValue = weightCoins.pools.reduce((m, v) => {
      return m.add(new Dec(v.value))
    }, new Dec(0))
    const coinsValue = weightCoins.coins.reduce((m, v) => {
      return m.add(new Dec(v.value))
    }, new Dec(0))
    const allocations = weightCoins.weights.reduce((m, v) => {
      return m.add(new Dec(v.allocation))
    }, new Dec(0))
    expect(Math.round(allocations)).toBe(1)

    // diff is zero or SUPER close to zero
    const diff = coinsValue.add(poolsValue).sub(new Dec(opts.total))
    expect(
      diff.isZero() || (diff.gt(new Dec(0)) && diff.lt(new Dec(0.0000001)))
    ).toBe(true)
    expect(weightCoins).toMatchSnapshot()
  },
  [
    {
      name: 'simple',
      ...bank,
      weights: [
        {
          weight: 5,
          denom: 'gamm/pool/3'
        },
        {
          weight: 5,
          denom: 'gamm/pool/1'
        },
        {
          weight: 5,
          poolId: '600'
        },
        {
          weight: 5,
          denom: 'gamm/pool/606'
        },
        {
          weight: 2,
          symbol: 'LUNA'
        },
        {
          weight: 10,
          symbol: 'UST'
        }
      ]
    },
    {
      name: 'coins-only',
      ...bank,
      weights: [
        {
          weight: 90,
          symbol: 'OSMO'
        },
        {
          weight: 2,
          symbol: 'LUNA'
        },
        {
          weight: 10,
          symbol: 'UST'
        }
      ]
    },
    {
      name: 'pools-only',
      ...bank,
      weights: [
        {
          weight: 5,
          denom: 'gamm/pool/1'
        },
        {
          weight: 5,
          poolId: '600'
        },
        {
          weight: 5,
          denom: 'gamm/pool/606'
        }
      ]
    },
    {
      name: 'single',
      ...bank,
      weights: [
        {
          weight: 5,
          denom: 'gamm/pool/3'
        }
      ]
    }
  ]
)
