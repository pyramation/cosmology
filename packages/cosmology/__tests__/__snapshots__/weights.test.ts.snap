// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`weights coins-only 1`] = `
Object {
  "coins": Array [
    Object {
      "amount": "935420168.0672268905070000",
      "denom": "uosmo",
      "displayAmount": "935.420168067226890507",
      "symbol": "OSMO",
      "value": "8512.323529411764703612",
    },
    Object {
      "amount": "2745069.5849374432750000",
      "denom": "ibc/0EF15DF2F02480ADE0BB6E85D9EBB5DAEA2836D3860E9F97F9AADE4F57A31AA0",
      "displayAmount": "2.745069584937443275",
      "symbol": "LUNA",
      "value": "189.162745098039216065",
    },
    Object {
      "amount": "945813725.4901960803230000",
      "denom": "ibc/BE1BB42D4BE3C30D50B68D7C41DB4DFCE9678E8EF8C539F6E6A9345048894FCC",
      "displayAmount": "945.813725490196080323",
      "symbol": "UST",
      "value": "945.813725490196080323",
    },
  ],
  "pools": Array [],
  "weights": Array [
    Object {
      "allocation": "0.882352941176470588",
      "denom": "uosmo",
      "name": "OSMO",
      "symbol": "OSMO",
      "type": "coin",
      "value": "8512.323529411764703612",
      "weight": 90,
    },
    Object {
      "allocation": "0.019607843137254902",
      "denom": "ibc/0EF15DF2F02480ADE0BB6E85D9EBB5DAEA2836D3860E9F97F9AADE4F57A31AA0",
      "name": "LUNA",
      "symbol": "LUNA",
      "type": "coin",
      "value": "189.162745098039216065",
      "weight": 2,
    },
    Object {
      "allocation": "0.098039215686274510",
      "denom": "ibc/BE1BB42D4BE3C30D50B68D7C41DB4DFCE9678E8EF8C539F6E6A9345048894FCC",
      "name": "UST",
      "symbol": "UST",
      "type": "coin",
      "value": "945.813725490196080323",
      "weight": 10,
    },
  ],
}
`;

exports[`weights pools-only 1`] = `
Object {
  "coins": Array [],
  "pools": Array [
    Object {
      "amount": "2376432056878314521733.181243142486789150",
      "coins": Array [
        Object {
          "amount": "43895258.8952588952590000",
          "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
          "displayAmount": "43.895258895258895259",
          "symbol": "ATOM",
          "value": "1607.883333333333333334",
        },
        Object {
          "amount": "176690476.1904761904760000",
          "denom": "uosmo",
          "displayAmount": "176.690476190476190476",
          "symbol": "OSMO",
          "value": "1607.883333333333333334",
        },
      ],
      "denom": "gamm/pool/1",
      "displayAmount": "-1",
      "name": "ATOM/OSMO",
      "value": "3215.766666666666666667",
    },
    Object {
      "amount": "16254142779185931.775473185986593475",
      "coins": Array [
        Object {
          "amount": "43895258.8952588952590000",
          "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
          "displayAmount": "43.895258895258895259",
          "symbol": "ATOM",
          "value": "1607.883333333333333334",
        },
        Object {
          "amount": "389317998.3857949959650000",
          "denom": "ibc/EA3E1640F9B1532AB129A571203A0B9F789A7F14BB66E350DCBFA18E1A1931F0",
          "displayAmount": "389.317998385794995965",
          "symbol": "CMDX",
          "value": "1607.883333333333333334",
        },
      ],
      "denom": "gamm/pool/600",
      "displayAmount": "-1",
      "name": "ATOM/CMDX",
      "value": "3215.766666666666666667",
    },
    Object {
      "amount": "207954394987295161840313.294065110450998271",
      "coins": Array [
        Object {
          "amount": "43895258.8952588952590000",
          "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
          "displayAmount": "43.895258895258895259",
          "symbol": "ATOM",
          "value": "1607.883333333333333334",
        },
        Object {
          "amount": "378168883787.7333935438500000",
          "denom": "ibc/B9E0A1A524E98BB407D3CED8720EFEFD186002F90C1B1B7964811DD0CCC12228",
          "displayAmount": "378168.883787733393543850",
          "symbol": "HUAHUA",
          "value": "1607.883333333333333334",
        },
      ],
      "denom": "gamm/pool/606",
      "displayAmount": "-1",
      "name": "ATOM/HUAHUA",
      "value": "3215.766666666666666667",
    },
  ],
  "weights": Array [
    Object {
      "allocation": "0.333333333333333333",
      "denom": "gamm/pool/1",
      "name": "ATOM/OSMO",
      "poolId": "1",
      "type": "pool",
      "value": "3215.766666666666663451",
      "weight": 5,
    },
    Object {
      "allocation": "0.333333333333333333",
      "denom": "gamm/pool/600",
      "name": "ATOM/CMDX",
      "poolId": "600",
      "type": "pool",
      "value": "3215.766666666666663451",
      "weight": 5,
    },
    Object {
      "allocation": "0.333333333333333333",
      "denom": "gamm/pool/606",
      "name": "ATOM/HUAHUA",
      "poolId": "606",
      "type": "pool",
      "value": "3215.766666666666663451",
      "weight": 5,
    },
  ],
}
`;

exports[`weights simple 1`] = `
Object {
  "coins": Array [
    Object {
      "amount": "8749909.3019881004210000",
      "denom": "ibc/0EF15DF2F02480ADE0BB6E85D9EBB5DAEA2836D3860E9F97F9AADE4F57A31AA0",
      "displayAmount": "8.749909301988100421",
      "symbol": "LUNA",
      "value": "602.956250000000000000",
    },
    Object {
      "amount": "3014781250.0000000000000000",
      "denom": "ibc/BE1BB42D4BE3C30D50B68D7C41DB4DFCE9678E8EF8C539F6E6A9345048894FCC",
      "displayAmount": "3014.781250000000000000",
      "symbol": "UST",
      "value": "3014.781250000000000000",
    },
  ],
  "pools": Array [
    Object {
      "amount": "10443364212422823981383.680393125000654635",
      "coins": Array [
        Object {
          "amount": "416406250.0000000000000000",
          "denom": "ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4",
          "displayAmount": "416.406250000000000000",
          "symbol": "AKT",
          "value": "753.695312500000000000",
        },
        Object {
          "amount": "82823660.7142857142860000",
          "denom": "uosmo",
          "displayAmount": "82.823660714285714286",
          "symbol": "OSMO",
          "value": "753.695312500000000000",
        },
      ],
      "denom": "gamm/pool/3",
      "displayAmount": "-1",
      "name": "AKT/OSMO",
      "value": "1507.390625000000000000",
    },
    Object {
      "amount": "1113952526661709932062.313239920996409930",
      "coins": Array [
        Object {
          "amount": "20575902.6071526071530000",
          "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
          "displayAmount": "20.575902607152607153",
          "symbol": "ATOM",
          "value": "753.695312500000000000",
        },
        Object {
          "amount": "82823660.7142857142860000",
          "denom": "uosmo",
          "displayAmount": "82.823660714285714286",
          "symbol": "OSMO",
          "value": "753.695312500000000000",
        },
      ],
      "denom": "gamm/pool/1",
      "displayAmount": "-1",
      "name": "ATOM/OSMO",
      "value": "1507.390625000000000000",
    },
    Object {
      "amount": "7619129427743405.519752266163153359",
      "coins": Array [
        Object {
          "amount": "20575902.6071526071530000",
          "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
          "displayAmount": "20.575902607152607153",
          "symbol": "ATOM",
          "value": "753.695312500000000000",
        },
        Object {
          "amount": "182492811.7433414043580000",
          "denom": "ibc/EA3E1640F9B1532AB129A571203A0B9F789A7F14BB66E350DCBFA18E1A1931F0",
          "displayAmount": "182.492811743341404358",
          "symbol": "CMDX",
          "value": "753.695312500000000000",
        },
      ],
      "denom": "gamm/pool/600",
      "displayAmount": "-1",
      "name": "ATOM/CMDX",
      "value": "1507.390625000000000000",
    },
    Object {
      "amount": "97478622650294607112636.752354254206427273",
      "coins": Array [
        Object {
          "amount": "20575902.6071526071530000",
          "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
          "displayAmount": "20.575902607152607153",
          "symbol": "ATOM",
          "value": "753.695312500000000000",
        },
        Object {
          "amount": "177266664275.5000282236060000",
          "denom": "ibc/B9E0A1A524E98BB407D3CED8720EFEFD186002F90C1B1B7964811DD0CCC12228",
          "displayAmount": "177266.664275500028223606",
          "symbol": "HUAHUA",
          "value": "753.695312500000000000",
        },
      ],
      "denom": "gamm/pool/606",
      "displayAmount": "-1",
      "name": "ATOM/HUAHUA",
      "value": "1507.390625000000000000",
    },
  ],
  "weights": Array [
    Object {
      "allocation": "0.156250000000000000",
      "denom": "gamm/pool/3",
      "name": "AKT/OSMO",
      "poolId": "3",
      "type": "pool",
      "value": "1507.390625000000000000",
      "weight": 5,
    },
    Object {
      "allocation": "0.156250000000000000",
      "denom": "gamm/pool/1",
      "name": "ATOM/OSMO",
      "poolId": "1",
      "type": "pool",
      "value": "1507.390625000000000000",
      "weight": 5,
    },
    Object {
      "allocation": "0.156250000000000000",
      "denom": "gamm/pool/600",
      "name": "ATOM/CMDX",
      "poolId": "600",
      "type": "pool",
      "value": "1507.390625000000000000",
      "weight": 5,
    },
    Object {
      "allocation": "0.156250000000000000",
      "denom": "gamm/pool/606",
      "name": "ATOM/HUAHUA",
      "poolId": "606",
      "type": "pool",
      "value": "1507.390625000000000000",
      "weight": 5,
    },
    Object {
      "allocation": "0.062500000000000000",
      "denom": "ibc/0EF15DF2F02480ADE0BB6E85D9EBB5DAEA2836D3860E9F97F9AADE4F57A31AA0",
      "name": "LUNA",
      "symbol": "LUNA",
      "type": "coin",
      "value": "602.956250000000000000",
      "weight": 2,
    },
    Object {
      "allocation": "0.312500000000000000",
      "denom": "ibc/BE1BB42D4BE3C30D50B68D7C41DB4DFCE9678E8EF8C539F6E6A9345048894FCC",
      "name": "UST",
      "symbol": "UST",
      "type": "coin",
      "value": "3014.781250000000000000",
      "weight": 10,
    },
  ],
}
`;

exports[`weights single 1`] = `
Object {
  "coins": Array [],
  "pools": Array [
    Object {
      "amount": "66837530959506073480855.554516000004189666",
      "coins": Array [
        Object {
          "amount": "2665000000.0000000000000000",
          "denom": "ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4",
          "displayAmount": "2665.000000000000000000",
          "symbol": "AKT",
          "value": "4823.650000000000000000",
        },
        Object {
          "amount": "530071428.5714285714290000",
          "denom": "uosmo",
          "displayAmount": "530.071428571428571429",
          "symbol": "OSMO",
          "value": "4823.650000000000000000",
        },
      ],
      "denom": "gamm/pool/3",
      "displayAmount": "-1",
      "name": "AKT/OSMO",
      "value": "9647.300000000000000000",
    },
  ],
  "weights": Array [
    Object {
      "allocation": "1.000000000000000000",
      "denom": "gamm/pool/3",
      "name": "AKT/OSMO",
      "poolId": "3",
      "type": "pool",
      "value": "9647.300000000000000000",
      "weight": 5,
    },
  ],
}
`;
